include/lbann/utils/dnn_lib/cudnn/convolution.hpp

```diff
@@ -0,0 +1,428 @@
+  El::AbstractDistMatrix<TensorDataType>& db)
+{
+  auto multisync = El::MakeMultiSync(gpu::get_sync_info(db),
+                                     gpu::get_sync_info(dy));
+  convolution_backward_bias(alpha_in, dyDesc, dy,
+                            beta_in, dbDesc, db,
+                            multisync);
+}
+
+template <typename TensorDataType, typename ScalarParameterType>
+void convolution_backward_filter(
+  ScalarParameterType const& alpha_in,
+  TensorDescriptor const& xDesc,
+  El::AbstractMatrix<TensorDataType> const& x,
+  TensorDescriptor const& dyDesc,
+  El::AbstractMatrix<TensorDataType> const& dy,
+  ConvolutionDescriptor const& convDesc,
+  bwd_conv_filter alg,
+  El::Matrix<TensorDataType, El::Device::GPU>& workSpace,
+  ScalarParameterType const& beta_in,
+  FilterDescriptor const& dwDesc,
+  El::AbstractDistMatrix<TensorDataType>& dw,
+  El::SyncInfo<El::Device::GPU> const& si)
+{
+  using LibScalingParamT = cudnn::ScalingParamType<TensorDataType>;
+  auto handle_manager = internal::make_default_handle_manager(si);
+  auto alpha = El::To<LibScalingParamT>(alpha_in);
+  auto beta = El::To<LibScalingParamT>(beta_in);
+  CHECK_CUDNN(cudnnConvolutionBackwardFilter(handle_manager.get(),
+                                      &alpha,
+                                      xDesc,
+                                      x.LockedBuffer(),
+                                      dyDesc,
+                                      dy.LockedBuffer(),
+                                      convDesc,
+                                      to_cudnn(alg),
+                                      workSpace.Buffer(),
+                                      workSpace.Height()*sizeof(TensorDataType),
+                                      &beta,
+                                      dwDesc,
+                                      dw.Buffer()));
+}
+
+template <typename TensorDataType, typename ScalarParameterType>
+void convolution_backward_filter(
+  ScalarParameterType const& alpha_in,
+  TensorDescriptor const& xDesc,
+  El::AbstractMatrix<TensorDataType> const& x,
+  TensorDescriptor const& dyDesc,
+  El::AbstractMatrix<TensorDataType> const& dy,
+  ConvolutionDescriptor const& convDesc,
+  bwd_conv_filter alg,
+  El::Matrix<TensorDataType, El::Device::GPU>& workSpace,
+  ScalarParameterType const& beta_in,
+  FilterDescriptor const& dwDesc,
+  El::AbstractDistMatrix<TensorDataType>& dw)
+{
+  auto multisync = El::MakeMultiSync(gpu::get_sync_info(dw),
+                                     gpu::get_sync_info(workSpace),
+                                     gpu::get_sync_info(dy),
+                                     gpu::get_sync_info(x));
+  convolution_backward_filter(alpha_in, xDesc, x, dyDesc, dy, convDesc, alg,
+                              workSpace, beta_in, dwDesc, dw, multisync);
+}
+
+template <typename TensorDataType, typename ScalarParameterType>
+void add_tensor(ScalarParameterType const& alpha_in,
```

Comment:
This is fine for now, but this seems to belong in the general cuDNN header rather than the conv header.

Reply:
here `include/lbann/utils/cudnn.hpp`?  If so, I'm currently working on a PR that will refactor `cudnn.hpp` and can include this.